src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,834 @@
+-- bits in the output bitvector). Return the resulting 'BV' along
+-- with its width.
+--
+-- >>> case bitsLE [True, False] of p -> (fstPair p, sndPair p)
+-- (2,BV 1)
+bitsLE :: [Bool] -> P.Pair NatRepr BV
+bitsLE [] = P.Pair (P.knownNat @0) (zero (P.knownNat @0))
+bitsLE (b:bs) = case bitsLE bs of
+  P.Pair w bv -> P.Pair w' bv'
+    where w' = w `addNat` (P.knownNat @1)
+          bv' = concat w (P.knownNat @1) bv (bool b)
+
+-- | Construct a 'BV' from a list of bytes, in big endian order (bytes
+-- with lower value index in the list are mapped to higher order bytes
+-- in the output bitvector). Return the resulting 'BV' along with its
+-- width.
+--
+-- >>> case bytesBE [2, 1] of p -> (fstPair p, sndPair p)
+-- (16,BV 513)
+bytesBE :: [Word8] -> P.Pair NatRepr BV
+bytesBE [] = P.Pair (P.knownNat @0) (zero (P.knownNat @0))
+bytesBE (b:bs) = case bytesBE bs of
+  P.Pair w bv ->  P.Pair w' bv'
+    where w' = (P.knownNat @8) `addNat` w
+          bv' = concat (P.knownNat @8) w (word8 b) bv
+
+-- | Construct a 'BV' from a list of bytes, in little endian order
+-- (bytes with lower value index in the list are mapped to lower order
+-- bytes in the output bitvector). Return the resulting 'BV' along
+-- with its width.
+--
+-- >>> case bytesLE [2, 1] of p -> (fstPair p, sndPair p)
+-- (16,BV 258)
+bytesLE :: [Word8] -> P.Pair NatRepr BV
+bytesLE [] = P.Pair (P.knownNat @0) (zero (P.knownNat @0))
+bytesLE (b:bs) = case bytesLE bs of
+  P.Pair w bv ->  P.Pair w' bv'
+    where w' = w `addNat` (P.knownNat @8)
+          bv' = concat w (P.knownNat @8) bv (word8 b)
 
--- | Unsigned interpretation of a bit vector as a (positive) Integer.
-bvIntegerU :: BitVector w -> Integer
-bvIntegerU (BV _ x) = x
+----------------------------------------
+-- BitVector -> Integer functions
 
--- | Signed interpretation of a bit vector as an Integer.
-bvIntegerS :: BitVector w -> Integer
-bvIntegerS bv = if bvTestBit bv (width - 1)
-                then bvIntegerU bv - (1 `shiftL` width)
-                else bvIntegerU bv
-  where width = bvWidth bv
+-- | Unsigned interpretation of a bitvector as a positive Integer.
+asUnsigned :: BV w -> Integer
+asUnsigned (BV x) = x
+
+-- | Signed interpretation of a bitvector as an Integer.
+asSigned :: (1 <= w) => NatRepr w -> BV w -> Integer
+asSigned w (BV x) =
+  -- NB, fromNatural is OK here because we maintain the invariant that
+  --  any existing BV value has a representable width
+  let wInt = fromNatural (natValue w) in
+  if B.testBit x (wInt - 1)
+  then x - B.bit wInt
+  else x
+
+-- | Unsigned interpretation of a bitvector as a Natural.
+asNatural :: BV w -> Natural
+asNatural = (fromInteger :: Integer -> Natural) . asUnsigned
+
+-- | Convert a bitvector to a list of bits, in big endian order
+-- (higher order bits in the bitvector are mapped to lower indices in
+-- the output list).
+--
+-- >>> asBitsBE (knownNat @5) (mkBV knownNat 0b1101)
+-- [False,True,True,False,True]
+asBitsBE :: NatRepr w -> BV w -> [Bool]
```

Comment:
We need to be a bit more careful about these enumerations, actually. We're doing the computations on `Natural`, so they might underflow.

Reply:
Same for asBitsLE

Reply:
Good catch, thanks.